### PR TITLE
feat: `near-primitives` Improve `combine_hash` performance by `40%`.

### PR DESCRIPTION
**Pull request by pmnoxx**
`combine_hash` in `core/primitives/src/merkle.rs` can be written in a cleaner way, that also avoid memory allocations:

```
pub fn combine_hash(hash1: &MerkleHash, hash2: &MerkleHash) -> MerkleHash {
    CryptoHash::hash_borsh(&(hash1, hash2))
 }
```

> I did some benchmarking.
> 
> Before any changes
> 
> ```
> test combine_hash_bench  ... bench:         143 ns/iter (+/- 1)
> ```
> 
> After not using Vec::new() and making arguments use references
> 
> ```
> test combine_hash_bench  ... bench:         102 ns/iter (+/- 1)
> ```

Btw, `MerkleHash` is the same thing as `CryptoHash`. Using references instead of doing memory copies, has some performance advantages.

If we don't change the signature and run the code like this
```
pub fn combine_hash2(hash1: MerkleHash, hash2: MerkleHash) -> MerkleHash {
    CryptoHash::hash_borsh(&(hash1, hash2))
}
```
We get performance result of:
```
test combine_hash_bench  ... bench:         123 ns/iter (+/- 1)
```

### CODE REVIEW

File: core/primitives/src/merkle.rs [hunk 1]
```diff
@@ -22,10 +22,8 @@ pub enum Direction {
     Right,
 }
 
-pub fn combine_hash(hash1: MerkleHash, hash2: MerkleHash) -> MerkleHash {
-    let mut combined: Vec<u8> = hash1.into();
-    combined.append(&mut hash2.into());
-    hash(&combined)
+pub fn combine_hash(hash1: &MerkleHash, hash2: &MerkleHash) -> MerkleHash {
```
Comment (mm-near) [inline core/primitives/src/merkle.rs:25]: did you verify that it generates the same results as before? can you add a small test here, that would compute the combine_hash - and compare to a fixed result (and make sure that this result is the same before and after your PR)

**Review by mm-near — CHANGES_REQUESTED** [review]

**Review by matklad — APPROVED** [review]
Yeah, would be great to confirm that some test fails if we actually mess up the impl.

File: core/primitives/src/merkle.rs (hunk 1 above)
Comment (pmnoxx) [reply core/primitives/src/merkle.rs:23]: @mm-near Yes, I did. I included a test in this PR.

**Comment by pmnoxx** [comment]
> Yeah, would be great to confirm that some test fails if we actually mess up the impl.

@matklad I added a test for this method, it works fine.

**Comment by pmnoxx** [comment]
Before
```
test combine_hash_bench  ... bench:         143 ns/iter (+/- 1)
```
After
```
test combine_hash_bench  ... bench:         102 ns/iter (+/- 1)
```

**Comment by pmnoxx** [comment]
> I did some benchmarking.
> 
> Before any changes
> 
> ```
> test combine_hash_bench  ... bench:         143 ns/iter (+/- 1)
> ```
> 
> After not using Vec::new() and making arguments use references
> 
> ```
> test combine_hash_bench  ... bench:         102 ns/iter (+/- 1)
> ```

If we don't change the signature and run the code like this
```
pub fn combine_hash2(hash1: MerkleHash, hash2: MerkleHash) -> MerkleHash {
    CryptoHash::hash_borsh(&(hash1, hash2))
}
```
We get performance result of:
```
test combine_hash_bench  ... bench:         123 ns/iter (+/- 1)
```

**Comment by pmnoxx** [comment]
@bowenwang1996 @mm-near It may be worth adding to readme, a not that 
`cargo test` needs `4-8gb of memory per thread`.

**Comment by pmnoxx** [comment]
```
@bowenwang1996 @mm-near @mzhangmzz: /chain/
```
@bowenwang1996 @mm-near @mzhangmzz Can you take a look?

**Comment by pmnoxx** [comment]
```
@bowenwang1996 @mm-near @mzhangmzz: chain/
```
@bowenwang1996 @mm-near @mzhangmzz Can you take a look?

**Comment by pmnoxx** [comment]
```
@bowenwang1996 @mm-near @mzhangmzz: chain/
```
@bowenwang1996 @mm-near @mzhangmzz Can you take a look?

**Review by bowenwang1996 — APPROVED** [review]
Thanks!

**Review by mzhangmzz — APPROVED** [review]

**Review by mm-near — APPROVED** [review]

**Pull request merged**